### PR TITLE
make payment processing more fault tolerant

### PR DESCRIPTION
**Pull request by nickpell**
http://manage.dimagi.com/default.asp?222722

:blowfish:

### CODE REVIEW

File: corehq/apps/accounting/payment_handlers.py [hunk 1]
```diff
@@ -102,10 +104,18 @@ def process_request(self, request):
...
+                    self.payment_method, None, amount
+                )
+                self.update_credits(payment_record)
+
+                charge = self.create_charge(amount, card=card, customer=customer)
 
-            charge = self.create_charge(amount, card=card, customer=customer)
+            payment_record.transaction_id = charge.id
+            payment_record.save()
             self.update_payment_information(billing_account)
```
Comment (snopoke) [inline corehq/apps/accounting/payment_handlers.py]: Is it intentional that these 3 lines are outside of the `atomic` block?

Comment (nickpell) [inline corehq/apps/accounting/payment_handlers.py]: yes - my intention is that the payment record is committed if and only if the charge is processed via stripe without error.  So I'm relying on the stripe command being at the end of the ```atomic``` block.  If any of those 3 lines error for unforeseen reasons, I don't want the payment record to be rolled back when the credit card has already been charged.

**Comment by nickpell** [comment]
@benrudolph / @proteusvacuum

File: corehq/apps/accounting/tests/test_stripe_payment.py [hunk 2]
```diff
@@ -35,6 +35,19 @@ def test_when_stripe_errors_no_payment_record_exists(self, mock_create):
 
         self.assertEqual(PaymentRecord.objects.count(), 0)
 
+    @patch.object(stripe.Charge, 'create')
+    @patch.object(PaymentRecord, 'create_record')
+    def test_when_create_record_fails_stripe_is_not_charged(self, mock_create_record, mock_create):
+        mock_create_record.side_effect = Exception
+
+        try:
+            self._call_process_request()
```
Comment (benrudolph) [inline corehq/apps/accounting/tests/test_stripe_payment.py]: nit: `self.assertRaises` is a nice way of doing this

**Comment by proteusvacuum** [comment]
:+1:

File: corehq/apps/accounting/tests/test_stripe_payment.py (hunk 2 above)
Comment (nickpell) [inline corehq/apps/accounting/tests/test_stripe_payment.py]: would've done that but the exception gets caught in ```process_request``` with the change I implemented. I was more interested in exposing the bad side-effect where our records get out of sync with what's been charged to stripe

**Comment by proteusvacuum** [comment]
@nickpell Out of curiosity, what are you using to organize imports?

**Comment by nickpell** [comment]
my usual approach is:

python standard library imports
django imports
3rd party imports
submodule imports
CCHQ imports

pycharm's "optimize imports" tool gets me most of the way

File: corehq/apps/accounting/tests/test_stripe_payment.py (hunk 2 above)
Comment (benrudolph) [inline corehq/apps/accounting/tests/test_stripe_payment.py]: :+1:

**Pull request merged**